### ISSUE TITLE
Using bulk under mysql stream

### ISSUE THREAD

**Issue by gjergjie** [elastic/elasticsearch-js#782]
Im trying to use this version of the lib within a mysql stream, like the above:

mysqlConnection.on('result', (data) => {
   client.bulk({
	refresh: true,
  	body: [ .... ]
   }, (err, res) => { console.... nothing outputs here});
});

So it seems it does not work within a stream, while I have used same scenario with an older version indexing data from MongoDB, and it used to work without any issue.....

Is there a way to handle this, or I should http request directly without using ES client?

Thanks!

**Comment by delvedor** [comment]
Hello!
Can you please write which library are you using (`@elastic/elasticsearch` or `elasticsearch`) and which version?
Can you also upload a full example to reproduce?

**Comment by gjergjie** [comment]
Its the current lib, I did a npm install @elastic/elasticsearch yesterday.. also the example its the above, so with a mysql stream, it does not seem to work.

**Comment by gjergjie** [comment]
I guess it should be an issue with mysql streaming rather than something with elastic client...... Trying to get to the bottom of this issue

**Comment by delvedor** [comment]
`@elastic/elasticsearch` do support streams.
In case of the bulk API, probably the issue is that the body is not well formed.
You can see [here](https://github.com/elastic/elasticsearch-js/blob/master/docs/examples/bulk.asciidoc) how to do it :)

**Comment by gjergjie** [comment]
Doesn't the malformed body throw errors?

**Comment by gjergjie** [comment]
Is there any example indexing data from mysql stream?

**Comment by delvedor** [comment]
> Doesn't the malformed body throw errors?

Nope, the client does not perform any kind of validation.
If the body is malformed, Elasticsearch will send back an error, and you will find it in the callback.

> Is there any example indexing data from mysql stream?

In JavaScript, not that I'm aware of.
But you can definitely do that with [Logstash](https://www.elastic.co/products/logstash)! :)

**Comment by gjergjie** [comment]
Cant use Logstash as Im doing ETL with the streaming results, so not easy to handle with Logstash.... Strange thing is that Im not getting any error, it's kinda stack there doing nothing.

**Comment by gjergjie** [comment]
Also, FYI Im checking the access logs in Easticsearch, as I have an nginx as a proxy, and the request does not get there at all.

**Comment by delvedor** [comment]
What is the value of `data` here?
```js
mysqlConnection.on('result', (data) => {
```

**Comment by gjergjie** [comment]
I have highWaterMark: 1, so it sends on data each record, and Im basically streaming emails from mysql and wanting to send to ES. 

Its just an object with email data, and on body param in the bulk method Im sending 

[
{"index" : { "_index": 'es-index', "_type": "type", "_id": md5(hash)}},
{... email data ...}
]

So trying with just one object, just to test if it works, but it gets stuck there. Also Im using the refresh param as true.

**Comment by delvedor** [comment]
Can you try to run `client.info(console.log)` ans see if the request works as expected?

**Comment by gjergjie** [comment]
Hmmm, it seems there is a memory issue..... at first it looks like its getting stuck, but if I wait for it, it's returning this:

<--- Last few GCs --->

[32640:0x3461060]    43986 ms: Mark-sweep 1416.3 (1483.8) -> 1415.2 (1484.8) MB, 1149.1 / 0.0 ms  allocation failure GC in old space requested
[32640:0x3461060]    45217 ms: Mark-sweep 1415.2 (1484.8) -> 1415.2 (1453.8) MB, 1231.1 / 0.0 ms  last resort GC in old space requested
[32640:0x3461060]    46288 ms: Mark-sweep 1415.2 (1453.8) -> 1415.2 (1453.8) MB, 1070.8 / 0.0 ms  last resort GC in old space requested


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x2524725258b9 <JSObject>
    1: addRequest [_http_agent.js:~138] [pc=0x65b93c38626](this=0x2a5ec2e17cb1 <Agent map = 0x3cf707e5f8a1>,req=0x20f910357351 <ClientRequest map = 0x2d513a3cb689>,options=0x20f9103575a1 <Object map = 0x2d513a3c6a39>,port=0x822fba822d1 <undefined>,localAddress=0x822fba822d1 <undefined>)
    2: arguments adaptor frame: 2->4
    3: new ClientRequest [_http_client.js:~84] [pc=0x65b93bbf246](this=0x...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
....

**Comment by gjergjie** [comment]
Strange, as I have done the exact same thing with MongoDB, which at the time, have had like more than 500TB of data, and the streaming to ES was working properly.... instead now on mysql there are not so many records, should be some issue on the stream management I guess.

**Comment by delvedor** [comment]
I'm sorry, but you are giving me a lot of unrelated data, and it's a little hard to understand what is happening.

Can you run the following code and write the result?
```js
'use strict'

const { Client } = require('@elastic/elasticsearch')
const client = new Client({ node: 'ELASTICSEARCH_URL' })

client.info(console.log)
```
Can you also write the version of Elasticsearch that you are using?

Thanks.

**Comment by gjergjie** [comment]
ES version is last one, 6.x installed it yesterday.

When I run the above everything works. But when I try to run it within the mysql stream, thats where the problem starts.

What I sent you is being generated from with the stream.

**Comment by delvedor** [comment]
I have several tests that verify that sending a stream is supported, so I would assume that the stream you are passing to the client has some issue.

Can you update your code like the following and paste here the result?
```diff
mysqlConnection.on('result', (data) => {
+ console.log(data)
  client.bulk({
    refresh: true,
    body: [ .... ]
  }, (err, res) => {
+    console.log(err | res)
  });
});
```

**Comment by gjergjie** [comment]
So thats the problem Im trying to understand, as the first console log works, then the second within the callback does not work.

**Comment by delvedor** [comment]
> Can you update your code like the following and **paste here the result**?

I'm sorry, but I have not enough data to understand what is going on.
Can you build a full example to reproduce it?

I recommend you to create a repository, you can use docker for MySql and Elasticsearch.
In the code, you should write the data into MySql and then run your logic for moving the data to Elasticsearch.

**Comment by gjergjie** [comment]
I gave you what I had.... You stating the exact problem I'm facing... so I can't get the error... something is stuck there.

Unfortunately this scenario is within a complex product, and I can not easily replicate it on some docker instance....

Anyway, Im going to deal with this on my own, thanks for your help!

**Comment by gjergjie** [comment]
Ok I got to the bottom of this, and the issue is on mysql driver, having an issue with managing the stream chunks over TCP, or implementation is not done properly in the connection pool example.

So the issue it's on mysql driver not on elasticsearch client.

I did it like this:

mysqlConnection.query( ... Query Here ... )
.stream({highWaterMark: 1}) // This is about how many objects/rows you want in a single chunk
.pipe(new stream.Transform({
      objectMode: true,
      transform: function (data, encoding, callback) {
           //Create bulk object here
           var body = [ ... ]
         
           _es.bulk({
		refresh: true,
		body: body
		}, (err, res) => { 
		    callback(); //Callback here to get into the next chunk
	   });
      }));

Hope it helps anyone facing same issue.
